In [1]:
import pandas as pd
import numpy as np

#### 1. Read the data set and replace dashes with 0 to make sure you can perform arithmetic operations on the data.

In [2]:
df=pd.read_csv('Laliga.csv')
df.head()

Unnamed: 0       Unnamed: 1 Unnamed: 2 Unnamed: 3   Unnamed: 4 Unnamed: 5  \
0        Pos             Team    Seasons     Points  GamesPlayed   GamesWon   
1          1      Real Madrid         86       4385         2762       1647   
2          2        Barcelona         86       4262         2762       1581   
3          3  Atletico Madrid         80       3442         2614       1241   
4          4         Valencia         82       3386         2664       1187   

   Unnamed: 6 Unnamed: 7 Unnamed: 8    Unnamed: 9 Unnamed: 10 Unnamed: 11  \
0  GamesDrawn  GamesLost   GoalsFor  GoalsAgainst    Champion   Runner-up   
1         552        563       5947          3140          33          23   
2         573        608       5900          3114          25          25   
3         598        775       4534          3309          10           8   
4         616        861       4398          3469           6           6   

  Unnamed: 12 Unnamed: 13 Unnamed: 14 Unnamed: 15 Unnamed: 16 Unnamed: 17  \
0       Third      Fourth       Fifth       Sixth           T       Debut   
1           8           8           3           4          79        1929   
2          12          12           4           6          83        1929   
3          16           9           7           6          56        1929   
4          10          11          10           7          50     1931-32   

     Unnamed: 18   Unnamed: 19  
0  Since/LastApp  BestPosition  
1           1929             1  
2           1929             1  
3        2002-03             1  
4        1987-88             1

In [3]:
df.columns=df.iloc[0]

In [4]:
df=df.drop(index=0)
df

0  Pos              Team Seasons Points GamesPlayed GamesWon GamesDrawn  \
1    1       Real Madrid      86   4385        2762     1647        552   
2    2         Barcelona      86   4262        2762     1581        573   
3    3   Atletico Madrid      80   3442        2614     1241        598   
4    4          Valencia      82   3386        2664     1187        616   
5    5   Athletic Bilbao      86   3368        2762     1209        633   
..  ..               ...     ...    ...         ...      ...        ...   
57  57             Xerez       1     34          38        8         10   
58  58            Condal       1     22          30        7          8   
59  59   Atletico Tetuan       1     19          30        7          5   
60  60  Cultural Leonesa       1     14          30        5          4   
61  61            Girona       1      -           -        -          -   

0  GamesLost GoalsFor GoalsAgainst Champion Runner-up Third Fourth Fifth  \
1        563     5947         3140       33        23     8      8     3   
2        608     5900         3114       25        25    12     12     4   
3        775     4534         3309       10         8    16      9     7   
4        861     4398         3469        6         6    10     11    10   
5        920     4631         3700        8         7    10      5     8   
..       ...      ...          ...      ...       ...   ...    ...   ...   
57        20       38           66        -         -     -      -     -   
58        15       37           57        -         -     -      -     -   
59        18       51           85        -         -     -      -     -   
60        21       34           65        -         -     -      -     -   
61         -        -            -        -         -     -      -     -   

0  Sixth   T    Debut Since/LastApp BestPosition  
1      4  79     1929          1929            1  
2      6  83     1929          1929            1  
3      6  56     1929       2002-03            1  
4      7  50  1931-32       1987-88            1  
5     10  49     1929          1929            1  
..   ...  ..      ...           ...          ...  
57     -   -  2009-10       2009-10           20  
58     -   -  1956-57       1956-57           16  
59     -   -  1951-52       1951-52           16  
60     -   -  1955-56       1955-56           15  
61     -   -  2017-18       2017-18            9  

[61 rows x 20 columns]

In [5]:
df= df.replace(to_replace='-', value = 0)
df

0  Pos              Team Seasons Points GamesPlayed GamesWon GamesDrawn  \
1    1       Real Madrid      86   4385        2762     1647        552   
2    2         Barcelona      86   4262        2762     1581        573   
3    3   Atletico Madrid      80   3442        2614     1241        598   
4    4          Valencia      82   3386        2664     1187        616   
5    5   Athletic Bilbao      86   3368        2762     1209        633   
..  ..               ...     ...    ...         ...      ...        ...   
57  57             Xerez       1     34          38        8         10   
58  58            Condal       1     22          30        7          8   
59  59   Atletico Tetuan       1     19          30        7          5   
60  60  Cultural Leonesa       1     14          30        5          4   
61  61            Girona       1      0           0        0          0   

0  GamesLost GoalsFor GoalsAgainst Champion Runner-up Third Fourth Fifth  \
1        563     5947         3140       33        23     8      8     3   
2        608     5900         3114       25        25    12     12     4   
3        775     4534         3309       10         8    16      9     7   
4        861     4398         3469        6         6    10     11    10   
5        920     4631         3700        8         7    10      5     8   
..       ...      ...          ...      ...       ...   ...    ...   ...   
57        20       38           66        0         0     0      0     0   
58        15       37           57        0         0     0      0     0   
59        18       51           85        0         0     0      0     0   
60        21       34           65        0         0     0      0     0   
61         0        0            0        0         0     0      0     0   

0  Sixth   T    Debut Since/LastApp BestPosition  
1      4  79     1929          1929            1  
2      6  83     1929          1929            1  
3      6  56     1929       2002-03            1  
4      7  50  1931-32       1987-88            1  
5     10  49     1929          1929            1  
..   ...  ..      ...           ...          ...  
57     0   0  2009-10       2009-10           20  
58     0   0  1956-57       1956-57           16  
59     0   0  1951-52       1951-52           16  
60     0   0  1955-56       1955-56           15  
61     0   0  2017-18       2017-18            9  

[61 rows x 20 columns]

Inference- we replaced all the '-' by 0 using the replace function.

#### 2. Print all the teams which have started playing between 1930-1980.

In [6]:
df['Debut'] = df['Debut'].str[:4]

In [7]:
df.Debut.unique()

array(['1929', '1931', '1934', '1939', '1932', '1941', '1948', '1944',
       '1935', '1949', '1933', '1960', '1951', '1998', '1977', '1959',
       '2004', '1961', '1940', '1930', '1963', '1974', '1943', '1987',
       '1991', '2007', '1962', '1994', '1978', '1971', '1999', '2014',
       '1990', '1947', '1996', '1995', '1945', '1953', '1979', '1950',
       '2016', '2009', '1956', '1955', '2017'], dtype=object)

In [8]:
df['Debut'] = df['Debut'].astype('int64')

In [9]:
r1=df.loc[(df.Debut>=1930)&(df.Debut<=1980),['Team','Debut']]
r1

0                  Team  Debut
4              Valencia   1931
6               Sevilla   1934
9              Zaragoza   1939
10           Real Betis   1932
11  Deportivo La Coruna   1941
12           Celta Vigo   1939
13           Valladolid   1948
15       Sporting Gijon   1944
16              Osasuna   1935
17               Malaga   1949
18               Oviedo   1933
19             Mallorca   1960
20           Las Palmas   1951
22              Granada   1941
23       Rayo Vallecano   1977
24                Elche   1959
26             Hercules   1935
27             Tenerife   1961
28               Murcia   1940
29               Alaves   1930
30              Levante   1963
31            Salamanca   1974
32             Sabadell   1943
33                Cadiz   1977
35            Castellon   1941
38              Cordoba   1962
40           Recreativo   1978
41            Burgos CF   1971
42           Pontevedra   1963
47            Gimnastic   1947
50             Alcoyano   1945
51                 Jaen   1953
53           AD Almeria   1979
55               Lleida   1950
58               Condal   1956
59      Atletico Tetuan   1951
60     Cultural Leonesa   1955

Inference- The above is the list of teams who started playing between 1930 and 1980.

#### 3. Print the list of teams which came Top 5 in terms of points 

In [10]:
df['Points']=df['Points'].astype('int64')
r2=df.sort_values(["Points"],ascending=False)

In [11]:
r2.loc[:,['Team','Points']].head()

0             Team  Points
1      Real Madrid    4385
2        Barcelona    4262
3  Atletico Madrid    3442
4         Valencia    3386
5  Athletic Bilbao    3368

Inference- Top 5 teams in terms of points.

#### 4. Write a function with name “Goal_diff_count” which should return all the teams with their Goal Differences. Using the same function, find the team which has maximum and minimum goal difference.
#### Goal_diff_count = GoalsFor - GoalsAgainst

In [12]:
df['GoalsFor']=df['GoalsFor'].astype('int64')
df['GoalsAgainst']=df['GoalsAgainst'].astype('int64')

In [13]:
def Goal_diff_count(Goals_For,Goals_Against):
    return Goals_For-Goals_Against

In [14]:
df['Goal_diff']=df.apply(lambda x: Goal_diff_count(x['GoalsFor'],x['GoalsAgainst']),axis=1)

In [15]:
df.head()

0 Pos             Team Seasons  Points GamesPlayed GamesWon GamesDrawn  \
1   1      Real Madrid      86    4385        2762     1647        552   
2   2        Barcelona      86    4262        2762     1581        573   
3   3  Atletico Madrid      80    3442        2614     1241        598   
4   4         Valencia      82    3386        2664     1187        616   
5   5  Athletic Bilbao      86    3368        2762     1209        633   

0 GamesLost  GoalsFor  GoalsAgainst  ... Runner-up Third Fourth Fifth Sixth  \
1       563      5947          3140  ...        23     8      8     3     4   
2       608      5900          3114  ...        25    12     12     4     6   
3       775      4534          3309  ...         8    16      9     7     6   
4       861      4398          3469  ...         6    10     11    10     7   
5       920      4631          3700  ...         7    10      5     8    10   

0   T Debut  Since/LastApp BestPosition Goal_diff  
1  79  1929           1929            1      2807  
2  83  1929           1929            1      2786  
3  56  1929        2002-03            1      1225  
4  50  1931        1987-88            1       929  
5  49  1929           1929            1       931  

[5 rows x 21 columns]

In [16]:
df[df['Goal_diff']==df['Goal_diff'].max()]

0 Pos         Team Seasons  Points GamesPlayed GamesWon GamesDrawn GamesLost  \
1   1  Real Madrid      86    4385        2762     1647        552       563   

0  GoalsFor  GoalsAgainst  ... Runner-up Third Fourth Fifth Sixth   T Debut  \
1      5947          3140  ...        23     8      8     3     4  79  1929   

0  Since/LastApp BestPosition Goal_diff  
1           1929            1      2807  

[1 rows x 21 columns]

Inference- Real Madrid is the team with maximum goal difference with 2807 goals.

In [17]:
df[df['Goal_diff']==df['Goal_diff'].min()]

0  Pos              Team Seasons  Points GamesPlayed GamesWon GamesDrawn  \
14  14  Racing Santander      44    1416        1428      453        336   

0  GamesLost  GoalsFor  GoalsAgainst  ... Runner-up Third Fourth Fifth Sixth  \
14       639      1843          2368  ...         1     1      2     0     1   

0   T Debut  Since/LastApp BestPosition Goal_diff  
14  5  1929        2011-12            2      -525  

[1 rows x 21 columns]

Inference- Racing Santander is the team with minimum goal difference with -525.

#### 5.  5. Create a new column with name “Winning Percent” and append it to the data set 

#### Percentage of Winning = (GamesWon / GamesPlayed)*100

#### If there are any numerical error, replace it with 0%


In [18]:
df['GamesWon']=df['GamesWon'].astype('int64')
df['GamesPlayed']=df['GamesPlayed'].astype('int64')

In [19]:
df['Winning Percent']=(df['GamesWon']/df['GamesPlayed'])*100

In [31]:
df['Winning Percent']=df['Winning Percent'].fillna(value=0)

In [32]:
df

0  Pos              Team Seasons  Points  GamesPlayed  GamesWon GamesDrawn  \
1    1       Real Madrid      86    4385         2762      1647        552   
2    2         Barcelona      86    4262         2762      1581        573   
3    3   Atletico Madrid      80    3442         2614      1241        598   
4    4          Valencia      82    3386         2664      1187        616   
5    5   Athletic Bilbao      86    3368         2762      1209        633   
..  ..               ...     ...     ...          ...       ...        ...   
57  57             Xerez       1      34           38         8         10   
58  58            Condal       1      22           30         7          8   
59  59   Atletico Tetuan       1      19           30         7          5   
60  60  Cultural Leonesa       1      14           30         5          4   
61  61            Girona       1       0            0         0          0   

0  GamesLost  GoalsFor  GoalsAgainst  ... Third Fourth Fifth Sixth   T Debut  \
1        563      5947          3140  ...     8      8     3     4  79  1929   
2        608      5900          3114  ...    12     12     4     6  83  1929   
3        775      4534          3309  ...    16      9     7     6  56  1929   
4        861      4398          3469  ...    10     11    10     7  50  1931   
5        920      4631          3700  ...    10      5     8    10  49  1929   
..       ...       ...           ...  ...   ...    ...   ...   ...  ..   ...   
57        20        38            66  ...     0      0     0     0   0  2009   
58        15        37            57  ...     0      0     0     0   0  1956   
59        18        51            85  ...     0      0     0     0   0  1951   
60        21        34            65  ...     0      0     0     0   0  1955   
61         0         0             0  ...     0      0     0     0   0  2017   

0  Since/LastApp  BestPosition Goal_diff Winning Percent  
1           1929             1      2807       59.630702  
2           1929             1      2786       57.241130  
3        2002-03             1      1225       47.475134  
4        1987-88             1       929       44.557057  
5           1929             1       931       43.772629  
..           ...           ...       ...             ...  
57       2009-10            20       -28       21.052632  
58       1956-57            16       -20       23.333333  
59       1951-52            16       -34       23.333333  
60       1955-56            15       -31       16.666667  
61       2017-18             9         0        0.000000  

[61 rows x 22 columns]

Inferece- Added a new column 'Winning Percent'. And replaced the null values with 0.

#### Print the top 5 teams which has the highest Winning percentage

In [33]:
r3=df.sort_values(["Winning Percent"],ascending=False)

In [34]:
r3.loc[:,['Team','Winning Percent']].head()

0             Team  Winning Percent
1      Real Madrid        59.630702
2        Barcelona        57.241130
3  Atletico Madrid        47.475134
4         Valencia        44.557057
5  Athletic Bilbao        43.772629

Inference- List if the top 5 teams with highest winning percent.

#### 6. Group teams based on their “Best position” and print the sum of their points for all positions

In [35]:
r4=df.groupby(by='BestPosition')['Points'].sum()

In [36]:
r4.sort_values(ascending=False)

BestPosition
1     27933
2      6904
4      6563
3      5221
6      2113
5      1884
7      1186
8      1134
12      511
10      450
11      445
17      266
9        96
16       81
19       81
14       71
20       34
15       14
Name: Points, dtype: int64

Inference- Sum of all the points based of the positions of the teams.

# Thank You 
### Dhairyav Shah.